framework/include/neml2/userobjects/NEML2Assembly.h

```diff
@@ -0,0 +1,69 @@
+//* This file is part of the MOOSE framework
+//* https://mooseframework.inl.gov
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#ifdef NEML2_ENABLED
+
+// MOOSE includes
+#include "ElementUserObject.h"
+
+#include "neml2/tensors/Tensor.h"
+
+/// This user object caches assebmly information from MOOSE.
+class NEML2Assembly : public ElementUserObject
+{
+public:
+  static InputParameters validParams();
+
+  NEML2Assembly(const InputParameters & parameters);
+
+  /// Number of active elements on this rank
+  int64_t numElem() const { return _nelem; }
+
+  /// Number of quadrature points per element
+  int64_t numQP() const { return _nqp; }
+
+  /**
+   * @brief Get the cached JxWxT for each element, each quadrature point
+   *
+   * The NEML2 tensor will have shape (nelem, nqp;), where nelem is the number of elements, and nqp
+   * is the number of quadrature points per element. The tensor is stored in the device memory.
+   *
+   * @warning The JxWxT value is only available after finalize() is called.
+   */
+  const neml2::Tensor & JxWxT() const { return _neml2_JxWxT; }
+
+  bool upToDate() const { return _up_to_date; }
+
+  /// Invalidate the cached assembly information
+  void invalidate() const;
```

Comment:
this doesn't seem like a method that should be `const`

Reply:
I see it called from `invalidateFEMContext` but that itself only seems to be called from the non-const `meshChanged`

Reply:
I'll try making that change and adding a test with refinement

Reply:
Oooooookaaaaay, looks like NEML2 and plain vanilla MOOSE segfault with adaptivity. We may need to update teh mass matrix (or its inverse) on mesh changes. I'm investigating this...